Spigot-Server-Patches/0565-Add-MonsterEggSpawnEvent.patch

```diff
@@ -0,0 +1,68 @@
+     }
+ 
+     @Nullable
+     public T spawnCreature(WorldServer worldserver, @Nullable NBTTagCompound nbttagcompound, @Nullable IChatBaseComponent ichatbasecomponent, @Nullable EntityHuman entityhuman, BlockPosition blockposition, EnumMobSpawn enummobspawn, boolean flag, boolean flag1) {
++        return this.spawnCreature(worldserver, nbttagcompound, ichatbasecomponent, entityhuman, blockposition, enummobspawn, flag, flag1, (ItemStack) null);
++    }
++
++    @Nullable
++    public T spawnCreature(WorldServer worldserver, @Nullable NBTTagCompound nbttagcompound, @Nullable IChatBaseComponent ichatbasecomponent, @Nullable EntityHuman entityhuman, BlockPosition blockposition, EnumMobSpawn enummobspawn, boolean flag, boolean flag1, ItemStack itemstack) {
+         // CraftBukkit start
+-        return this.spawnCreature(worldserver, nbttagcompound, ichatbasecomponent, entityhuman, blockposition, enummobspawn, flag, flag1, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SPAWNER_EGG);
++        return this.spawnCreature(worldserver, nbttagcompound, ichatbasecomponent, entityhuman, blockposition, enummobspawn, flag, flag1, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SPAWNER_EGG, itemstack);
+     }
+ 
+     @Nullable
+     public T spawnCreature(WorldServer worldserver, @Nullable NBTTagCompound nbttagcompound, @Nullable IChatBaseComponent ichatbasecomponent, @Nullable EntityHuman entityhuman, BlockPosition blockposition, EnumMobSpawn enummobspawn, boolean flag, boolean flag1, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason spawnReason) {
+-        T t0 = this.createCreature(worldserver, nbttagcompound, ichatbasecomponent, entityhuman, blockposition, enummobspawn, flag, flag1);
++        return this.spawnCreature(worldserver, nbttagcompound, ichatbasecomponent, entityhuman, blockposition, enummobspawn, flag, flag1, spawnReason, null);
++    }
++
++    @Nullable
++    public T spawnCreature(WorldServer worldserver, @Nullable NBTTagCompound nbttagcompound, @Nullable IChatBaseComponent ichatbasecomponent, @Nullable EntityHuman entityhuman, BlockPosition blockposition, EnumMobSpawn enummobspawn, boolean flag, boolean flag1, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason spawnReason, ItemStack itemstack) {
++        // Paper end
++        T t0 = this.createCreature(worldserver, nbttagcompound, ichatbasecomponent, entityhuman, blockposition, enummobspawn, flag, flag1); // Paper - diff on change
++
++        // Paper start - MonsterEggSpawnEvent
++        if (spawnReason == CreatureSpawnEvent.SpawnReason.SPAWNER_EGG && itemstack != null && t0 != null) {
++            MonsterEggSpawnEvent event = new MonsterEggSpawnEvent(entityhuman != null ? (org.bukkit.entity.Player) entityhuman.getBukkitEntity() : null, (org.bukkit.entity.LivingEntity) t0.getBukkitEntity(), itemstack.asBukkitCopy());
```

Comment:
This handling seems flawed if something hits something which adds the entity to the world before something calls setEntity, etc, should remove the original entity ref too

Reply:
Creating one seems to be sufficient. Why spawning an entity in world if you remove it the same second again on cancel? 
So this should be fine:

1. create entity
2. throw monster egg event
3. if not cancelled add it to world which results into a normal spawn event

Shouldn't something like this work?

`
 @Nullable
    public Entity spawnCreature(WorldServer worldserver, @Nullable ItemStack itemstack, @Nullable EntityHuman entityhuman, BlockPosition blockposition, EnumMobSpawn enummobspawn, boolean flag, boolean flag1) {
        // Paper start
        if (itemstack != null && entityhuman != null) {
            T t0 = this.createCreature(worldserver, itemstack.getTag(), itemstack != null && itemstack.hasName() ? itemstack.getName() : null, entityhuman, blockposition, enummobspawn, flag, flag1);
            if (t0 != null) {
                io.papermc.paper.event.entity.MonsterEggSpawnEvent monsterEggSpawnEvent = new io.papermc.paper.event.entity.MonsterEggSpawnEvent(
                    entityhuman.getBukkitEntity(), t0.getBukkitEntity(), itemstack.asBukkitCopy());
                if (monsterEggSpawnEvent.callEventIfRegistered()) {
                    Entity t1 = ((org.bukkit.craftbukkit.entity.CraftEntity)monsterEggSpawnEvent.getEntity()).getHandle();
                    worldserver.addAllEntities(t1, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SPAWNER_EGG);
                    return !t1.dead ? t1 : null; // Don't return an entity when CreatureSpawnEvent is canceled
                    // CraftBukkit end
                }
            }
        }
        // Paper end
        return this.spawnCreature(worldserver, itemstack == null ? null : itemstack.getTag(), itemstack != null && itemstack.hasName() ? itemstack.getName() : null, entityhuman, blockposition, enummobspawn, flag, flag1);
    }
`

Reply:
Because there are many interactions which plugin can do easily which can add an entity to the world before expected, e.g. teleporting the entity will add it to the world

Reply:
I will add on. That I am willing to update this PR to be mergeable in the upcoming week due to my familiarity with it. I have a Spring break this next week starting Friday! 😄